README.md
README

# grid-correl-atcor
Set of scripts for spatially-variable radiometric normalization of satellite imagery and preparation of Sentinel-2 L2A imagery for use in GIS. The scripts named L2A_*.sh work specifically with Sentinel-2 Level-2A imagery, while the scripts i.grid.correl.atcor.py and r.buff.cloudmask.py can be used on raster data from various sources. Please note: **This code repository is not yet complete. The scripts presented are minimally tested and under development.** 

# The scripts
## i.grid.correl.atcor.py
Provides the core functionality, ie. the radiometric normalization of single band of a satellite image based on reference image. It works within the [GRASS GIS](https:/grass.osgeo.org) 7.x session. When run without arguments, it provides graphical user interface. 
### Installation and initialization
For starting it in GUI mode from the GRASS GIS menu anytime, and also for initial registration of the script within your GRASS GIS, save it somewhere (preferably in some directory you intend for storing also other third-party GRASS scripts) and run it using the GRASS menu *File / Launch Script*. From now on, it is added into that GRASS GIS user search PATH, so you can use its name on the commnad-line, for example to do a loop over all bands of a satellite image imported into GRASS working mapset.
### Basic principle
To achieve its purpose, that is to do the radiometric normalization in spatially-variable manner, it processes the input raster per tiles. In every tile, the correlation coeffitient *r* between the input and reference image is computed for every tile, and if it is better than the minimum and number of valid (ie. pseudo/invariant) pixels is over the minimum, linear regression slope *b* and intercept *a* between reference and input image tiles are computed. The slope and intercept are then interpolated over the whole area of the image. The slope and intercept rasters are then used to compute corrected raster band.
Before the linear regression computation, the image should be masked, so that only the so called pseudo-invariant area pixels are used for the computation. It is user's responsibility to provide the required masks (but the other scripts in the set are here to help with that).
### Synopsis
```
Spatially variable correlation based radiometric normalization.

Usage:
 i.grid.correl.atcor.py [-kv] input=string reference=string output=name
   [masks=string[,string,...]] [gridsize=value] [pixels=value]
   [minr=value] [regression=string] [interpolation=string]
   [lambda_i=value] [--overwrite] [--help] [--verbose] [--quiet] [--ui]

Flags:
  -k   Keep temporary files created during operation.
  -v   Verbose processing information.

Parameters:
          input   Select the band to be corrected.
      reference   Select the reference band.
         output   Select name of output corrected band.
          masks   Select the raster(s) to mask out invalid/changing pixels.
       gridsize   Approx. grid tile size in map units (6000 in m means box 6x6 km).
                  default: 6000
         pixels   Minimal number of valid pixels in tile.
                  default: 100
           minr   Minimal correlation coefficient R to accept.
                  default: 0.85
     regression   Regression method: theil_sen - TheilSen regression, orthogonal - orthogonal regression, least_sq - ordinary least squares.
                  values:theil_sen,orthogonal,least_sq
                  default: theil_sen
  interpolation   Select interpolation method (v.surf.bspline).
                  values:bilinear,bicubic
                  default: bicubic
       lambda_i   Tykhonov regularization parameter (v.surf.bspline)
                  default: 0.1
```
*Older releases of i.grid.correl.atcor.py with some additional documentation can be found at [this Dropbox link](https://www.dropbox.com/s/st5b4p5nkmn8t3k/i.grid.correl.atcor.html?dl=0). (No sign-up required, just close the pop up - but due to changes in Dropbox site you now need to download the html file and open it in browser for it to be rendered if you are not signed in.)*
* * *
## L2A_grass_atcor.sh
This Bash script allows using the script *i.grid.correl.atcor.py* on Sentinel-2 imagery without starting grass manually and process the whole set of bands of L2A Sentinel-2 product in one run. It needs set of input files created using *L2A_vrt-img.sh* script and the reference image already stored in PERMANENT mapset of the location used. Before first use make sure to edit user settings directly within the code.
### Synopsis
```
Usage:  
     L2A_grass_atcor.sh [-a "<atcor parameters>"] <Input_reflective_bands_file.img>
To get help:
     L2A_grass_atcor.sh -h
To get version info:  
     L2A_grass_atcor.sh -v

You can also simply drop the input file to process on the script icon in GUI (no explicit progress indication in the GUI, but you can monitor the log files). The otput file and logs will be generated where the input is stored.
 
DESCRIPTION

L2A_grass_atcor.sh - A wrapper shell script for creating grass temporary mapset 
		in existing location, importing/creating necessary files and
		running i.grid.correl.atcor.py. Supposes input files created with
		L2A_vrt-img and the reference file already stored in PERMANENT 
		mapset of the loaction. This instance invoked as: 
		./L2A_grass_atcor.sh
 
PARAMETERS

-a "<atcor parameters>"
--atcor_parms "<atcor parameters>"
		(optional) Parameters that will be passed to the 
		i.grid.correl.atcor.py script. If used, these replace the whole 
		set of defaults specified within this script (currently: 
		gridsize=4000 minr=0.88 pixels=300 regression=orthogonal). 
		Must be passed as single string enclosed in double quotes. For 
		help run 'i.grid.correl.atcor.py --help' within a GRASS GIS 
		session.

```
* * *
## L2A_vrt-img.sh
Script to take zip file with Sentinel-2 L2A SAFE T33UWR imagery, unzip it and create .vrt and .img files for all resolution image bands for that tile. Also works on already unpacked .SAFE directory. Additionaly, the script creates 20m resolution water and cloud+shade masks and MNDWI, NDMI and NDVI indices.
Files generated by this script are used by *L2A_grass_atcor.sh*, but are also suitable for general use in GIS software, like [QGIS](https://qgis.osgeo.org). For that reason the script creates also some files not used by L2A_grass_atcor.sh, like the 10m and 60m multiband .img files.
Before first use make sure to edit user settings directly within the code.

The multiband .img and .vrt files MSI band order:

10m (file \<TILENAME\>_\<TIMESTAMP\>_10m.img): 02, 03, 04, 08

20m (file \<TILENAME\>_\<TIMESTAMP\>_20m.img): 02, 03, 04, 05, 06, 07, 11, 12, 8A

60m (file \<TILENAME\>_\<TIMESTAMP\>_60m.img): 01, 02, 03, 04, 05, 06, 07, 09, 11, 12, 8A

### Synopsis
```
Usage:                L2A_vrt-img.sh [-o] <Input SAFE format directory or zip archive>
To get help:          L2A_vrt-img.sh -h
To get version info:  L2A_vrt-img.sh -v

Runtime switches:
-o, --overwrite		Overwrite existing files

You can also drop the input file/directory to process on the script icon in GUI. The files will be generated where the input is stored. Existing output files creation is skipped, unless the --overwrite switch is used.
 
L2A_vrt-img.sh - Script to take zip file with Sentinel-2 L2A SAFE T33UWR imagery, unzip it and create .VRT and .IMG files for all resolution image bands for that tile (for other tile than T33UWR - edit the USER VARIABLES in the script). 
          This instance invoked as /home/tom/scripts/L2A_vrt-img.sh
```
**Please note, [issue 2](https://github.com/tomas4/grid-correl-atcor/issues/2) have been closed, but it brought change in values reported by QGIS - it now reports values of reflectance, numbers less than 1, while previous version reported quantized values, numbers generally in thousands. This may affect your workflows. For example if you have prepared color styles for the files, you have to rework them, or the multiband files will generally look as solid black.**
***
## r.buff.cloudmask.py
Script to despeckle and buffer cloud mask derived from SCL classification (or other cloud mask containing artifacts in the form of misclassified small few pixel clouds or small holes in them). It works within the [GRASS GIS](https:/grass.osgeo.org) 7.x session. The buffering is there also to mask out areas in close vicinity of detected clouds, where usually are present thin clouds not detected properly and strong neigborhood effects (parasite light reflected off cloud edge etc.). The script is needed by L2A_grass_atcor.sh.
See *i.grid.correl.atcor.py* for installation instructions.

### Synopsis
```
Script to grow zero value (cloudy) regions in a cloud mask band.

Usage:
 r.buff.cloudmask.py input=string [clmask=name] [output=name]
   [buffsize=value] [circlesize=value] [--overwrite] [--help] [--verbose]
   [--quiet] [--ui]

Flags:

Parameters:
       input   Select the mask raster to be buffered.
      clmask   Name of output cleared mask raster. Leave empty and input name with suffix _clean<size> will be used.
      output   Name of output cleared and buffered raster. Leave empty and input name with suffix _buff<size> will be used.
    buffsize   Buffer size in meters.
               default: 300
  circlesize   Size of moving window circular area to filter out few pixel-sized clouds and holes. The value must be an odd number >= 3.
               default: 9
```
***
## L1C_fmask.sh
Simple wrapper script for [FMASK](https://github.com/gersl/fmask) algorithm to create alternative cloud mask to that created by *L2A_vrt_img.sh* from level-2 scene classification of Sentinel-2 imagery. Note that you need FMASK4.x installation (tested with FMASK 4.3) and have to edit the user settings within the *L1C_fmask.sh* file. Also note that you need Level-1C Sentinel-2 image, not Level-2A in this case. In many cases the FMASK 4.3 based cloudmask is higher quality than L2A SCL based cloudmask. To use the resulting cloudmask by the *L2A_grid_atcor.sh* script, make backup of the *L2A_vrt_img.sh* created cloudmask file and rename the FMASK based cloudmask exactly as L2A SCL based cloudmask was named.

### Synopsis
```
Usage:                L1C_fmask.sh [-o] <Input SAFE format directory or zip archive>
To get help:          L1C_fmask.sh -h
To get version info:  L1C_fmask.sh -v

Runtime switches:
-o, --overwrite		Overwrite existing files

You can also drop the input file/directory to process on the script icon in GUI. The files will be generated where the input is stored. Existing output files creation is skipped, unless the --overwrite switch is used.
 
L1C_fmask.sh - Script to take .SAFE folder (or .zip file containing it) with Sentinel-2 L1C imagery and create cloud mask using FMASK. Resulting raster would be named like this: T33UWR_20210306T100029_cloud_fmask_20m.img. T33UWR is granule (tile) name currently set for processing, for other granule edit the user settings within the script. STANDARD NAMING OF FILES DOWNLOADED FROM COPERNICUS OPEN DATA HUB IS SUPPOSED. 
          This instance invoked as /home/tom/scripts/L1C_fmask.sh

```



Changelog.md
Changelog

# Changelog of i.grid.correl.atcor.py releases
Previous releases below were published via Dropbox (see download links below). Following releases would be released in github project https://github.com/tomas4/grid-correl-atcor together with complementary scripts. Scripts here on Github undergone substantial changes since latest Dropbox release. The chages could be tracked via Github logs now.

**DATE: 2015-07-23
VERSION: 0.92b**

Download: https://www.dropbox.com/s/g8v2edcdihr4hsu/i.grid.correl.atcor_0.92b.tar.gz?dl=0

**Beta version, compatible with GRASS 7
New features:** 

- OrtInformation messages now contain the input file name and mapset path in the first few lines.
- hogonal regression - new method of regression, made default.

**Fixes:**

- Removed the stddev_mean and combined methods of regression, these were wrong.

**Other changes:**

- The regression parameter now has options least_sq and orthogonal, for "ordinary least squares" nad "orthogonal regression" methods of linear regression.
- .py suffix in the command file name reintroduced, since it is needed on Windows.
- (internal) Some code streamlining (mostly simplification of second stage mask creation, removal of old comments and commented out pieces of code).

**Known isssues:**

- The script does not work for the author in current versions of WinGRASS 7 (tested with 7.0.0 and 7.0.1 RC1, please report if it does for you) due to remaining issues in these Windows versions of GRASS. Works fine on Linux and probaly OSX. This is also true for the previous script version 0.91b.

**DATE: 2015-04-08
VERSION: 0.91b**

Download: https://www.dropbox.com/s/fc3s4b697g3vnhf/i.grid.correl.atcor_0.91b.zip?dl=0

Beta version, first release compatible with GRASS 7
**New features:**

- GRASS 7 compatibility (no longer compatible with GRASS 6.4)
- New method to compute the regression line (based on stdev and mean of input bands)

**New options:**

- regression - chooses method of computing regression parameters (least_squares,mean_stdev,combined).

**Fixes:**

- The WxGUI option to add created map to layer tree now works.
- The resulting corrected band now should cover the whole area of input bands overlap within region bounds (no rounnded corners anymore).
- Other changes:
- minR parameter default relaxed to 0.8.
- Masks parameter moved to optional (advanced) parameters.
- Prints number of tiles used for regression computation.
- (internal) r.sum (removed from GRASS 7) replaced by r.univar for counting valid pixels within tile.
- Removed .py suffix in the command name.

**DATE: released 2015-02-12 as 0.9a, re-released with updated documentation 2015-02-23 as 0.9b**

VERSION: 0.9b, 0.9a
Downloads:
https://www.dropbox.com/s/cqykxrjja0zo9yc/i.grid.correl.atcor_0.9b.zip?dl=0
https://www.dropbox.com/s/2urin9zw7k3clze/i.grid.correl.atcor_0.9a.zip?dl=0

- Internal alpha version, later first public release together with short html documentation with workflow description.
- New features: Interpolation methods bilinear, bicubic (using v.surf.bspline) and their parmeters, IDW interpolation method dropped. Switch (k) to keep temporary files.
- New options: interpolation, lamda_i, k
- Other changes: Advanced parameters section in the GUI, with interpolation parameters. Some default values of parameters altered.

**DATE: 2014-01-20
VERSION: 0.7a**

- Unfinished alpha, internal.
- New features: Raster(s) to mask out invalid/changing pixels. Removal of temporary files.
- New options: masks

**DATE: 2014-01-16
VERSION: 0.6a**

- Unfinished alpha, internal. First version producing an corrected image. IDW interpolation not good.
- New features: Minimal correlation coefficient R to accept, storing regression parameters to vector point file attributes, interpolation of the regession parameters into rasters (using v.surf.idw), computation of the output map.
- New options: minR

**DATE: 2014-01-14
VERSION: 0.5a**

- Unfinished alpha, internal. Incomplete.
- Implemented: Grid size and creation, parsing the tiles, minimal number of valid pixels in grid tile, computation of regression parameters (just reported on command line), storing and restoring region.
- Options: input, reference, output, gridsize, pixels

**Development started: 2013-11-22**


L2A_modelapp.sh
#!/bin/bash

# L2A_modelapp - Noninteractive script. USAGE:
# 1. EDIT VARIABLES IN <parameters file>
# 2. Run the script within directory with the files created with L2A_vrt-img
#
# TIP: Store this as a template, create and edit a copy in the source directory, then run!
# 
#
# Have fun!
# Tomas IV. (Tomas Brunclik, brunclik@atlas.cz)



## Versioning settings
SCRIPT_NAME=$(basename $0)
SCRIPT_VERSION="0.1 (2019-06-10)" 
SCRIPT_YEAR="2019"
SCRIPT_AUTHOR="Tomas Brunclik"

## Version info (Only if the -v or --version option is the first cmdline parameter)
if [ "-v" = "${1}" -o "--SCRIPT_VERSION" = "${1}" ]
then
    cat <<!
$SCRIPT_NAME $SCRIPT_VERSION (c) $SCRIPT_AUTHOR $SCRIPT_YEAR
Use under terms of GNU General Public License.
See http://www.gnu.org/licenses/gpl.html or file gpl.txt for details.
!
    exit
fi
 
## Help (Only if the -h or --help option is the first cmdline parameter, or the number of parameters is not 2.)
if [ "-h" = "${1}" -o "--help" = "${1}" -o $# -ne 1 ]
then
    cat <<!
 
Usage:                $SCRIPT_NAME <parameters file>
To get help:          $SCRIPT_NAME -h
To get version info:  $SCRIPT_NAME -v

$SCRIPT_NAME - NON-INTERACTIVE script to apply a water quality model on Sentinel-2 satellite level-2 data preprocessed with L2A_vrt-img script. EDIT THE <parameters file> FIRST. Output would be created in current directory. 



          This instance invoked as ${0}
 
!
    exit
fi


## Functions
#function name () { list; } [redirection]


## Main program

# Check existence of the input file with parameters
if [ ! -e ./${1} ]
then
    echo "The file ${1} not found. It should be text file with model parameters in current directory, run '$SCRIPT_NAME -h' for more info."
    exit 1
fi

# Check presence of tools needed
command -v gdal_calc.py >/dev/null 2>&1 || { echo >&2 "I require gdal_calc.py but it's not installed. Please install gdal tools (gdal-bin package or similar).  Aborting."; exit 1; }
# command -v basename >/dev/null 2>&1 || { echo >&2 "I require basename but it's not installed. Please install core linux utilities (coreutils package or similar).  Aborting."; exit 1; }
# command -v dirname >/dev/null 2>&1 || { echo >&2 "I require dirname but it's not installed. Please install core linux utilities (coreutils package or similar).  Aborting."; exit 1; }

# Source the parameters
source ./${1}

# Compute it 
# Distinguish between formulas with 5, 4, 3 or 2 operands
if [ "$E" ]; then
    gdal_calc.py -A "$A" --A_band=$ABAND -B "$B" --B_band=$BBAND -C "$C" --C_band=$CBAND -D "$D" --D_band=$DBAND -E "$E" --E_band=$EBAND  --outfile="$OUTPUT" --type='Float32' --calc="$FORMULA"
elif [ "$D" ]; then
	gdal_calc.py -A "$A" --A_band=$ABAND -B "$B" --B_band=$BBAND -C "$C" --C_band=$CBAND -D "$D" --D_band=$DBAND --outfile="$OUTPUT" --type='Float32' --calc="$FORMULA"
elif [ "$C" ]; then
	gdal_calc.py -A "$A" --A_band=$ABAND -B "$B" --B_band=$BBAND -C "$C" --C_band=$CBAND --outfile="$OUTPUT" --type='Float32' --calc="$FORMULA"
elif [ "$B" ]; then
	gdal_calc.py -A "$A" --A_band=$ABAND -B "$B" --B_band=$BBAND --outfile="$OUTPUT" --type='Float32' --calc="$FORMULA"
fi



L2A_grass_atcor.sh
#!/bin/bash

# L2A_grass_atcor - A wrapper shell script for creating grass temporary mapset in  existing location, importing/creating necessary files and running i.grid.correl.atcor.py. Supposes input files created with L2A_vrt-img and the reference files already stored in PERMANENT mapset of the loaction. 
#
# For newest version visit http://
#
# CREDITS:
# 
#
# Have fun!
# Tomas IV. (Tomas Brunclik, brunclik@atlas.cz)

#################### USER VARIABLES (edit to your needs) ####################
#Location full path
THELOC="/home/tom/GISdata/grass/utm33n"
#REFERENCE maps in PERMANENT mapset
# Basename of reference image bands. For example if the band names are reference.1, reference.2, ..., reference.9, then it should be REFBASE="reference"
REFBASE="s2_20220719_20m"
# Cloudmask of the reference image, may also mask out snow if needed. If the image is clear without any clouds and snow, just give it empty string value, ie. REFCLOUDMASK=""
#REFCLOUDMASK="s2_20220719_20m_1mask"
REFCLOUDMASK="" 
REFSCL="s2_20220719_20m_scl"
REFNDMI="s2_20220719_20m_ndmi"
# The GRASS command
GRASSCMD=/usr/bin/grass
#GRASSCMD=grass78
#Directory with python scripts i.grid.correl.atcor.py and r.buff.cloudmask.py
SCRIPTDIR="/home/tom/Dropbox/grass-moje/grass-run"
#i.grid.correl.atcor.py parameters (multiple space separated parameters and/or flags except input/output; leave empty for default values, use the commented example below as a guide for modifications)
#ATCOR_PARMS=""
ATCOR_PARMS="gridsize=4000 minr=0.88 pixels=300 regression=orthogonal"
##NDMI difference mask parameters
# max NDMI difference to pass the mask (to filter out change in vegetation cover and moisture content)
MAXNDMIDIFF="0.1"
# max absolute NDMI to pass the mask (to allow only dry/impervious surfaces which are generally more stable over time). Set it to 1.0 to effectively disable filtering high NDMI (may be needed in regions with full vegetation cover).
MAXNDMI="0.15"
#Suffixes of INPUT files created with L2A_vrt-img (usually editing not needed)
INPUTSX="_20m.img"
CLOUDSX="_cloud_mask_20m.img"
WATERSX="_water_mask_20m.img"
NDMISX="_ndmi_20m.img"
SCLSX="_SCL_20m.vrt"
# Output file suffix. It must be different, than above suffixes. 
# corr4or - 20220719 reference, orthogonal regression, results straight in reflectance as the reference.
OUTPUTSX="_20m_corr4or.img"
#############################################################################


## Versioning settings
SCRIPT_NAME=$(basename $0)
SCRIPT_VERSION="0.5 (2023-04-20)" 
SCRIPT_YEAR="2019"
SCRIPT_AUTHOR="Tomas Brunclik"

## Version info (Only if the -v or --version option is the first cmdline parameter)
if [ "-v" = "${1}" -o "--SCRIPT_VERSION" = "${1}" ]
then
    cat <<!
$SCRIPT_NAME $SCRIPT_VERSION (c) $SCRIPT_AUTHOR $SCRIPT_YEAR
Use under terms of GNU General Public License.
See http://www.gnu.org/licenses/gpl.html or file gpl.txt for details.
!
    exit
fi
 
## Help (Only if the -h or --help option is the first cmdline parameter)
if [ "-h" = "${1}" -o "--help" = "${1}" ]
then
    cat <<!
 
Usage:  
     $SCRIPT_NAME [-a "<atcor parameters>"] <Input_reflective_bands_file.img>
To get help:
     $SCRIPT_NAME -h
To get version info:  
     $SCRIPT_NAME -v

You can also simply drop the input file to process on the script icon in GUI (no explicit progress indication in the GUI, but you can monitor the log files). The otput file and logs will be generated where the input is stored.
 
DESCRIPTION

$SCRIPT_NAME - A wrapper shell script for creating grass temporary mapset 
		in existing location, importing/creating necessary files and
		running i.grid.correl.atcor.py. Supposes input files created with
		L2A_vrt-img and the reference file already stored in PERMANENT 
		mapset of the loaction. This instance invoked as: 
		${0}
 
PARAMETERS

-a "<atcor parameters>"
--atcor_parms "<atcor parameters>"
		(optional) Parameters that will be passed to the 
		i.grid.correl.atcor.py script. If used, these replace the whole 
		set of defaults specified within this script (currently: 
		$ATCOR_PARMS). 
		Must be passed as single string enclosed in double quotes. For 
		help run 'i.grid.correl.atcor.py --help' within a GRASS GIS 
		session.
!
    exit
fi

## Functions
#function name () { list; } [redirection]


## Main program

### Parse command-line parameters ############
while [[ "$#" -gt 0 ]]; do case $1 in

  -a|--atcor_parms) 
		ATCOR_PARMS="$2"
		shift
		;;

  *)	
  		if [ -e "$1" ]
  		then
  			INPATHFILE="$1"
  		else
  			echo "ERROR: Can not find input reflective bands file $1 supplied. Make sure it has correct path."; 
  			exit 1
  		fi
  		;;
  		
  esac; shift 
done


### CHECKS ##############################

# get separate dirname and filename
INDNAME="$(dirname $INPATHFILE)"
INFNAME="${INPATHFILE#"${INDNAME}/"}"

# set the base name
L2ABASE="${INFNAME%"$INPUTSX"}"

# Check if output file exists
echo "Checking output file..."
echo "Output directory: $INDNAME"
echo "Output file: ${L2ABASE}${OUTPUTSX}"
if [ -e ${INDNAME}${L2ABASE}${OUTPUTSX} ]
then
  read -t 20 -p "WARNING: Output file already exists and will be OWERWRITTEN. Press N to exit, Y or anything else to continue and overwrite the file in the end." 
  case $REPLY in
	  [yY]*)
	    echo "Going ahead."
	    ;;
	  [nN]*)
	    echo "Exitting."
	    exit 0
	    ;;
	  *) echo "Going ahead."
	     ;;
  esac
else
  echo "OK"
fi
echo

# Check presence of reference maps in PERMANENT mapset
echo "Checking presence of reference maps in PERMANENT mapset..."
echo "Location: ${THELOC}"
for band in 1 2 3 4 5 6 7 8 9 ; do
  $GRASSCMD ${THELOC}/PERMANENT --exec g.findfile element=cell file="${REFBASE}.$band" > /dev/null 2>&1 || { echo "ERROR: map ${REFBASE}.$band not found in PERMANENT mapset. Please check REFERENCE MAPS in USER SETTINGS and in your PERMANENT."; exit 1; }
  echo "${REFBASE}.$band OK"
done
echo

# Check presence of reference auxiliary maps in PERMANENT mapset
echo "Checking presence of reference maps in PERMANENT mapset..."
for map in $REFCLOUDMASK $REFSCL $REFNDMI; do
  $GRASSCMD ${THELOC}/PERMANENT --exec g.findfile element=cell file="$map" > /dev/null 2>&1 || { echo "ERROR: map $map not found in PERMANENT mapset. Please check REFERENCE MAPS in USER SETTINGS and in your PERMANENT."; exit 1; }
  echo "$map OK"
done
echo

# Check presence of all input files
echo "Checking presence of all input files in ${INDNAME}..."
for suffix in $INPUTSX $CLOUDSX $WATERSX $NDMISX $SCLSX; do
  if [ ! -e "${INDNAME}/${L2ABASE}${suffix}" ]
  then
    echo "ERROR: File ${L2ABASE}${suffix} not found. Make sure all the input files were generated correctly by L2A_vrt-img.sh script and check input files suffixes in USER SETTINGS."
    exit 1
  fi
  echo "${L2ABASE}${suffix} OK" 
done
echo

# Check presence of scripts
echo "Checking presence of scripts in ${SCRIPTDIR}..."
[ -e "${SCRIPTDIR}/i.grid.correl.atcor.py" ] || { echo "ERROR: script i.grid.correl.atcor.py not found. Please check SCRIPTDIR path in USER SETTINGS and content of the directory specified."; exit 1; }
echo i.grid.correl.atcor.py OK
[ -e "${SCRIPTDIR}/r.buff.cloudmask.py" ] || { echo "ERROR: script r.buff.cloudmask.py not found. Please check SCRIPTDIR path in USER SETTINGS and content of the directory specified."; exit 1; }
echo r.buff.cloudmask.py OK
echo
#########################################



### PREPARE GRASS MAPSET ################
#create mapset
MAPSETNAME="tmp$$"
echo "Creating mapset $MAPSETNAME in location $THELOC"
$GRASSCMD -c -e ${THELOC}/${MAPSETNAME}

#link/import files 
# New 2021-02: Do not use original ${L2ABASE} within the GRASS mapset, because it can contain unsupported characters. Use "input" and "output" instead. Means to do the replace down to the ## CREATE & EXPORT OUTPUT ## part (inclusive).  
# Also stripping the trailing filetype suffix from the ${suffix}, like this: _cloud_mask_20m.img --> _cloud_mask_20m (i.e. with ${suffix%.*}).
# Also reverted back to full import (r.in.gdal) instead of linking (r.external), since there were problems with ndmi file with nodata values defined as very high number. The parameters of both commands are identical. ###DEVIDEA: This makes it extremely easy to make the link/import method user-selectable.
echo "Importing maps to mapset $MAPSETNAME"
for suffix in $INPUTSX $CLOUDSX $WATERSX $NDMISX $SCLSX; do
  $GRASSCMD ${THELOC}/${MAPSETNAME} --exec r.in.gdal -oe input="${INDNAME}/${L2ABASE}${suffix}" output="input${suffix%.*}" > /dev/null 2>&1 || { echo "ERROR: map ${L2ABASE}$suffix import failed."; exit 1; }
done
echo

#set region and null mask based on imported SCL file
echo "Setting region"
$GRASSCMD ${THELOC}/$MAPSETNAME --exec g.region raster="input${SCLSX%.*}" > /dev/null 2>&1 || { echo "ERROR: Setting region failed."; exit 1; }
echo "Masking no-data areas"
$GRASSCMD ${THELOC}/$MAPSETNAME --exec r.mask -i raster="input${SCLSX%.*}" maskcats=0  > /dev/null 2>&1 2>&1 || echo "WARNING: Creating MASK for bands no-data areas failed." #Missing mask is non-fatal. It only prolongs the atcor processing for Sentinel-2 granules containing substantial nodata areas and makes these areas zero-value.
#########################################



### CREATE CHANGE AND FEATURE MASKS #####
echo "Creating SCL difference mask"
$GRASSCMD ${THELOC}/$MAPSETNAME --exec r.mapcalc expression="scldiffmask = $REFSCL == input${SCLSX%.*}" > /dev/null 2>&1 || { echo "ERROR: Creating SCL difference mask failed."; exit 1; }
echo "Creating NDMI difference mask"
# NEW 2020-02: Using smoothing of both ndmi rasters first to mitigate image noise influence and to get true PIFs as features (areas) as opposed to scattered pixels.
$GRASSCMD ${THELOC}/$MAPSETNAME --exec r.neighbors input="input${NDMISX%.*}" output=ndmi_in_smooth method=average size=3
$GRASSCMD ${THELOC}/$MAPSETNAME --exec r.neighbors input="$REFNDMI" output=ndmi_ref_smooth method=average size=3
# NEW 2020-02: Limit the mask to areas with low moisture content (NDMI<0.15) 
$GRASSCMD ${THELOC}/$MAPSETNAME --exec r.mapcalc expression="ndmidiffmask = ( (ndmi_in_smooth - ndmi_ref_smooth) > -$MAXNDMIDIFF && (ndmi_in_smooth - ndmi_ref_smooth) < $MAXNDMIDIFF && (input${NDMISX%.*} < $MAXNDMI) && ($REFNDMI < $MAXNDMI) )" > /dev/null 2>&1 || { echo "ERROR: Creating NDMI difference mask failed."; exit 1; }
echo "Buffering and cleaning cloud mask"
$GRASSCMD ${THELOC}/$MAPSETNAME --exec python ${SCRIPTDIR}/r.buff.cloudmask.py input="input${CLOUDSX%.*}" buffsize=300 output="input${CLOUDSX%.*}_buff300" > /dev/null 2>&1 || { echo "ERROR: Buffering and cleaning cloud mask failed."; exit 1; }
#########################################



### CREATE & EXPORT OUTPUT ##############
read -p "Starting the i.grid.correl.atcor.py for all bands (press Ctrl-C to abort now)" -t 5
echo
#run i.grid.correl.atcor.py
time for i in 1 2 3 4 5 6 7 8 9
do 
    LOG=${L2ABASE}${OUTPUTSX}.${i}_$(date +%s).log
    if [ -z "$REFCLOUDMASK" ]; then 
      $GRASSCMD ${THELOC}/$MAPSETNAME --exec python ${SCRIPTDIR}/i.grid.correl.atcor.py --overwrite -k $ATCOR_PARMS input=input${INPUTSX%.*}.$i reference=${REFBASE}.$i output=input${OUTPUTSX%.*}.$i masks=input${CLOUDSX%.*},input${CLOUDSX%.*}_buff300,ndmidiffmask,scldiffmask 2>&1 | tee tmplog_$$ || exit 1
    else
      $GRASSCMD ${THELOC}/$MAPSETNAME --exec python ${SCRIPTDIR}/i.grid.correl.atcor.py --overwrite -k $ATCOR_PARMS input=input${INPUTSX%.*}.$i reference=${REFBASE}.$i output=input${OUTPUTSX%.*}.$i masks=${REFCLOUDMASK},input${CLOUDSX%.*},input${CLOUDSX%.*}_buff300,ndmidiffmask,scldiffmask 2>&1 | tee tmplog_$$ || exit 1
    fi
    #Tidy up the log
    #The sed filter removes terminal codes for progress percents, grep empty lines with spaces
    cat tmplog_$$ | sed 's/\x1B\[[0-9;]\+[A-Za-z]//g' | grep "\S" > $LOG
    rm tmplog_$$
done 

# create group of images, export
#group
$GRASSCMD ${THELOC}/$MAPSETNAME --exec i.group group=input${OUTPUTSX%.*} input=input${OUTPUTSX%.*}.1,input${OUTPUTSX%.*}.2,input${OUTPUTSX%.*}.3,input${OUTPUTSX%.*}.4,input${OUTPUTSX%.*}.5,input${OUTPUTSX%.*}.6,input${OUTPUTSX%.*}.7,input${OUTPUTSX%.*}.8,input${OUTPUTSX%.*}.9 > /dev/null 2>&1 || { echo "ERROR: Creating image group input${OUTPUTSX%.*} failed."; exit 1; }
#export
$GRASSCMD ${THELOC}/$MAPSETNAME --exec r.out.gdal -f --overwrite input=input${OUTPUTSX%.*} output=${INDNAME}/${L2ABASE}${OUTPUTSX} format=HFA type=Float32 createopt=COMPRESSED=YES
#########################################



### REMOVE MAPSET #######################
# Rather check the $MAPSETNAME variable is not empty before deleting (because if it was, the command would delete the whole location!)
#echo "Removing mapset $MAPSETNAME"
#[ "$MAPSETNAME" ] && rm -rf "${THELOC}/$MAPSETNAME"
echo
echo "The script used temporay GRASS GIS mapset $MAPSETNAME in the location ${THELOC}. It is now safe to delete that mapset (but you may want to inspect it first for the purpose of fine tuning input parameters, quality checking, etc...)."
echo
if [ -e ${INDNAME}/${L2ABASE}${OUTPUTSX} ]
then
    echo "Output file ${INDNAME}/${L2ABASE}${OUTPUTSX} created."
else
    echo "Something went wrog. Check the messages above and the logs."
fi
echo "All done."
#########################################



